server/src/pages/shot/text-tool.js

```diff
@@ -0,0 +1,266 @@
+    }
+    const styles = window.getComputedStyle(this.textInput.current);
+    const FONT_SIZE = parseInt(styles["font-size"], 10);
+    const x = this.state.left + parseFloat(styles["padding-left"]);
+
+    // For large and small font increase y position by BUFFER_VERTICAL pixel
+    // to fix vertical shift seen when rendering text on canvas
+    let BUFFER_VERTICAL = 0;
+
+    if (FONT_SIZE === 36) {
+      BUFFER_VERTICAL = 1;
+    } else if (FONT_SIZE === 16) {
+      BUFFER_VERTICAL = 0.5;
+    }
+
+    const y = this.state.top + BUFFER_VERTICAL + parseFloat(styles.height) / 2;
+
+    const textCanvas = document.createElement("canvas");
+    textCanvas.width = this.props.baseCanvas.width;
+    textCanvas.height = this.props.baseCanvas.height;
+    const drawingContext = textCanvas.getContext("2d");
+
+    drawingContext.scale(this.props.canvasPixelRatio, this.props.canvasPixelRatio);
+    drawingContext.textBaseline = "middle";
+    drawingContext.fillStyle = styles.backgroundColor;
+    drawingContext.fillRect(this.state.left,
+                            this.state.top,
+                            this.textInput.current.clientWidth,
+                            this.textInput.current.clientHeight);
+    drawingContext.fillStyle = styles.color;
+    drawingContext.font = `${FONT_WEIGHT} ${FONT_SIZE}px ${FONT_STYLE}`;
+    drawingContext.fillText(this.textInput.current.textContent, x, y);
+
+    const textSelection = new Selection(this.state.left,
+                                        this.state.top,
+                                        this.state.left + this.textInput.current.clientWidth,
+                                        this.state.top + this.textInput.current.clientHeight);
+
+    if (this.props.confirmTextHandler) {
+      this.props.confirmTextHandler(textSelection, textCanvas);
+    }
+    sendEvent("confirm-text", "text-toolbar");
+  }
+
+  onClickCancel(e) {
+    if (this.props.cancelTextHandler) {
+      this.props.cancelTextHandler();
+    }
+    sendEvent("cancel-text", "text-toolbar");
+  }
+
+  onKeyDown(e) {
+    // Allow backspace to delete text
+    const key = e.keyCode || e.charCode;
+    if ( key === BACKSPACE_KEYCODE || key === DELETE_KEYCODE ) {
+      return;
+    }
+    // Stop entering text if text content is longer than image width and
+    // prevent new line insertion if user press Enter key
+    if (parseFloat(this.textInput.current.clientWidth + 2 * DRAG_DIV_PADDING) > this.canvasCssWidth ||
+        key === ENTER_KEYCODE) {
+      e.preventDefault();
+    }
+  }
+
+  onKeyUp(e) {
+    // Update left state to left position of text input
+    const rectInput = e.target.getBoundingClientRect();
+    const rectCanvas = this.props.baseCanvas.getBoundingClientRect();
+    const WIDTH_DIFF = this.textInput.current.clientWidth - previousTextInputWidth;
+    this.setState({left: rectInput.left - rectCanvas.left - WIDTH_DIFF / 2});
+    previousTextInputWidth = this.textInput.current.clientWidth;
```

Comment:
I think this would be better in onkeydown for the times when a user hold down a key to insert or hold down backspace to delete text.

Reply:
this.textInput.current.clientWidth inside keydown doesn't change and  gets updated to clientWidth with new inserted character only onkeyUp

Reply:
By this do you mean moving just line 224 in onKeyDown, that could work